[POS][Custom payment UI] – Failed payment error handling

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->
### Failed payment error handling 

Closes: #12827
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Design specs: TfaZ4LUkEwEGrxfnEFzvJj-fi-2710_173633

⚠️ Don't merge until the base branch is merged into `trunk`.

The intent of this PR is to display the "Payment failed" full-screen state in case the payment fails. The error screen should allow two actions:
1. Retry payment
2. Exit order

### Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Go to Woo POS (Menu more > Point of Sale)
2. Add items to cart and click "Checkout"
3. Ensure reader is connected
4. Tap the card to the reader and simulate failed payment. An easy way to do that is to tap a real card when running a debug build.
5. Verify that "Payment failed" screen is shown
6. Test "Try another payment method" and "Exit order" buttons

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

- [x] Payment failed screen should match design specs
- [x] "Try another payment method" button action should redirect back to Checkout and allow tapping card again
- [ ] "Exit order" button should redirect to Cart, and clear the cart, allowing to make a new order
- [ ] Failed payment state restoration across config changes

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->
* Samsung Tab S9 FE, android 14
* Tested "failed payment" state in different scenarios
* Tested "Try another payment method" button action
* Tested "Exit order" button action
* Tested state restoration with "don't keep activities" enabled

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/7da5ad9d-88e8-40f1-b917-5fc77e7842e2

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->12827

samiuelson committed:
Show full screen "Payment processing" state

samiuelson committed:
Clean up code

samiuelson committed:
Add preview

samiuelson committed:
Clean up code

samiuelson committed:
Improve preview

samiuelson committed:
Create basic payment failed screen

samiuelson committed:
Show payment failed screen

dangermattic commented:
<!--
  1 Error: This PR is tagged with <code>s...
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">This PR is tagged with <code>status: do not merge</code> label(s).</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

samiuelson committed:
Update tests

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Tune up payment failed composable

samiuelson committed:
Tune up payment failed composable

samiuelson committed:
Tune up payment failed composable

samiuelson committed:
Rename states

samiuelson committed:
Ensure failed payment state is always full screen

samiuelson committed:
Implement failed payment "retry" action

samiuelson committed:
Implement failed payment "exit order" action

samiuelson committed:
Move debug Totals payment state to the top

samiuelson committed:
Add home VM test:

give home screen is at checkout, when exit order clicked after failed payment, then should show cart

samiuelson committed:
Add home VM test:

give home screen is at checkout, when payment processing started, then should show full screen totals state

samiuelson committed:
Add home VM test:

give home screen is at checkout, processing payment, when payment fails, then should show full screen totals state

samiuelson committed:
Add home VM test:

give home screen is at checkout, failed payment, when retry payment clicked, then should show cart with totals

samiuelson committed:
Tweak heights of totals boxes

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/payment/processing/WooPosTotalsPaymentProcessingScreen.kt:
```diff
@@ -0,0 +1,48 @@
+import com.woocommerce.android.ui.woopos.home.totals.WooPosTotalsViewState
+
+@Composable
+fun WooPosPaymentProcessingScreen(
```
💡 The final designs are not implemented but are coming in the next PR.

samiuelson committed:
Fix typos in test method names

samiuelson committed:
Add Cart VM test:

given non-empty cart, when card payment is aborted, then should clear the cart

samiuelson committed:
Add Totals VM test:

given order draft created and reader connected, when card tapped, should show payment processing screen

samiuelson committed:
Add Totals VM test:

given payment failed, when retry clicked, then should retry

samiuelson committed:
Add Totals VM test:

given payment failed, when retry clicked, then should retry

samiuelson committed:
Add Totals VM test:

given payment failed, when exit order clicked, then should inform home about the situation

samiuelson committed:
Ensure Cart is cleared in case state is restored

samiuelson committed:
Clean up code

samiuelson committed:
Satisfy detekt's complaints

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsScreen.kt:
```diff
@@ -116,18 +133,25 @@ private fun TotalsLoaded(
+            when {
+                error != null -> TotalsError(modifier = Modifier, error = error)
+                else -> {
+                    Text(
```
💡 This is still a "debug" Text showing the current payment state. The "Tap card" animation will be implemented in the next PR.

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt:
```diff
@@ -145,6 +151,20 @@ class WooPosTotalsViewModel @Inject constructor(
+            WooPosTotalsUIEvent.RetryFailedTransactionClicked -> viewModelScope.launch {
+                cancelPaymentAction()
+                childrenToParentEventSender.sendToParent(ChildToParentEvent.RetryFailedPaymentClicked)
+                if (order == null) {
```
💡 After "Try another payment method" button is clicked, the `order` here will be null in case the Activity/VM gets destroyed. I think it's fine to redirect to Cart to start the order again in such case.

kidinov replied:
Why not just get the order from the database?

E.g.

https://github.com/woocommerce/woocommerce-android/blob/e7bd63a005a83860afc85b2ca311cd6577b18dbe/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/payment/receipt/WooPosTotalsPaymentReceiptRepository.kt#L48

samiuelson replied:
Good idea. Done – cf8d030cfaccd8576fc3ab79fe9df91a77e915da

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/WooPosTheme.kt:
```diff
@@ -195,7 +196,8 @@ private val DarkCustomColors = CustomColors(
     paymentSuccessIcon = WooPosColors.darkCustomColorsHomeBackground,
-    homeBackground = WooPosColors.darkCustomColorsHomeBackground
+    homeBackground = WooPosColors.darkCustomColorsHomeBackground,
+    paymentProcessingBackground = Color(0xFF533582),
```
We need to use palette colors - `WooPurple70`

samiuelson replied:
Fixed – 51eb2b2fd35b382f4086ae00de83e90c7588583d

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/common/composeui/WooPosTheme.kt:
```diff
@@ -208,7 +210,8 @@ private val LightCustomColors = CustomColors(
     paymentSuccessIcon = Color.White,
-    homeBackground = WooPosColors.Gray0
+    homeBackground = WooPosColors.Gray0,
+    paymentProcessingBackground = Color(0xFF533582),
```
Just validating that dark and light mode has the same color?

samiuelson replied:
Couldn't find anything better in Figma, but as far as I can see it's the same in both modes on iOS as well. I'll check with the designer.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt:
```diff
@@ -199,17 +220,57 @@ class WooPosTotalsViewModel @Inject constructor(
-                        )
-                    childrenToParentEventSender.sendToParent(ChildToParentEvent.OrderSuccessfullyPaid)
+
+                when (paymentState) {
```
Maybe not in this PR, but it would be nice to extract all these payment handling related code to some sort of helper to keep the VM lean

samiuelson replied:
Good point. I'll do it in the next PR, once all the logic handling different payment states is complete (e.g. displaying card reader hints).

kidinov reviewed (commented):
LGTM! Cool that the animation worked pretty out of the box

I found a weird behavior though. I have a product which was 125 USD prices, I changed it to 0.20 cents to emulate an error. Not it shows the error first, but when I try again it passed 

https://github.com/user-attachments/assets/3cc68459-4f49-4e94-b93a-32fffaca0d64

---

If I touch the card reader a couple times and leave the flow, it'll deliver the error state a few times

https://github.com/user-attachments/assets/10051d4f-8883-428d-b1fd-cfafa8c64666

---

I didn't get this on video but I saw 125 USD prices on the final screen once. Seems liek some caching issue

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt:
```diff
@@ -199,17 +220,57 @@ class WooPosTotalsViewModel @Inject constructor(
         }
     }
 
+    private fun buildPaymentFailedState(): PaymentFailed = PaymentFailed(
```
Why don't we use string id from `PaymentFlowError` to show an error text that helps to resolve the issue?

samiuelson replied:
I initially wanted to do that in a separate PR, but it turned out to be very simple, so I did it in this commit – 8b4f5f7

samiuelson committed:
Use WooPosColors for `paymentProcessingBackground`

samiuelson committed:
Use WooPosColors for `totalsBackground`

samiuelson committed:
Replace hardcoded error message with real one

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Get order from db instead of caching in memory

samiuelson committed:
Improve failed payment "retry" handling

samiuelson committed:
Satisfy detekt's complaints

samiuelson committed:
Update string value

Exit order -> Go back to checkout

samiuelson committed:
Handle PaymentCollecting payment state during retry

samiuelson committed:
Handle `PaymentCollecting` payment state received during retry

samiuelson committed:
Clean up code

samiuelson commented:
Thanks for the review, @kidinov! 

> I found a weird behavior though. I have a product which was 125 USD prices, I changed it to 0.20 cents to emulate an error. Not it shows the error first, but when I try again it passed

I've analyzed the iOS implementation (p1733476692009279-slack-C070SJRA8DP), and I updated the error handling logic in 5b12dd8 and 6c2a7e7 accordingly. In nutshell, the actions behind buttons are modified:
1. **Try payment again** is shown in case the failed payment state provides `onRetry` lambda. Clicking that button results in `onRetry` being called, and retrying payment (not even necessarily requiring the card to be tapped again).
2. **Try another payment method** is shown in case `onRetry` lambda is null. Clicking that button cancels current payment intent, redirects back to Checkout and starts payment flow again.
3. **Go back to checkout** is shown only in pair with **Try payment again**. Clicking that button has the same effect as **Try another payment method**.

Now, in case of $0.20 purchase, the error should persist when you click "Try payment again". However, the transaction is successful when you restart the payment collection from scratch by clicking "Go back to checkout". This seems to be the intended behavior and happens in IPP as well (see the video below) – the payment fails the first time, also consecutive retries fail, however, if we cancel that payment intent and repeat the payment action from scratch, it goes through.

https://github.com/user-attachments/assets/71afc9a9-aa2b-4d24-8ab9-76dbc7725b96

> If I touch the card reader a couple times and leave the flow, it'll deliver the error state a few times

This should be fixed now.

> I didn't get this on video but I saw 125 USD prices on the final screen once. Seems liek some caching issue.

This is weird. I tried the same scenario—updating the price of a product to $0.20—but I didn't manage to reproduce it. The total value comes directly from the server's order creation response though, so it could have been a backend issue (total in `dataState.value` is not set from any device data). Are you still able to reproduce it?

Let me know what you think!

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt:
```diff
@@ -193,23 +230,92 @@ class WooPosTotalsViewModel @Inject constructor(
+                when (paymentState) {
+                    is CardReaderPaymentState.CollectingPayment,
+                    is CardReaderPaymentState.LoadingData ->
+                        handlePaymentState(paymentState as CardReaderPaymentState)
```
No cast needed here

samiuelson replied:
Done - 8b21898

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/totals/WooPosTotalsViewModel.kt:
```diff
@@ -193,23 +230,92 @@ class WooPosTotalsViewModel @Inject constructor(
+                paymentStateText = paymentState.javaClass.simpleName
+            )
+        } else {
+            val order = totalsRepository.getOrderById(dataState.value.orderId)
```
Are you sure that in both cases where you get the order, it won't be better just to crash if it's not in the DB? Or is there a legit case when its null?

samiuelson replied:
Yeah, probably that makes more sense to crash. Done - 8b21898

kidinov reviewed (approved):
@samiuelson 👋

I retested, looks good! I left a couple of comments too, please take a look

> This seems to be the intended behavior and happens in IPP as well (see the video below) – the payment fails the first time, also consecutive retries fail, however, if we cancel that payment intent and repeat the payment action from scratch, it goes through.

You think this is like that because it's a test mode?

What's weird is that now it looks like that we don't show proper message why the payment fails for the amounts lower than minimal amount. Like something wrong either on the backend side and we get the wrong response or on our side

samiuelson committed:
Merge branch `custom-payment-ui-6-use-payment-controller-in-pos`

samiuelson commented:
Thanks for review, @kidinov!

> > This seems to be the intended behavior and happens in IPP as well (see the video below) – the payment fails the first time, also consecutive retries fail, however, if we cancel that payment intent and repeat the payment action from scratch, it goes through.
> 
> You think this is like that because it's a test mode?
> 
> What's weird is that now it looks like that we don't show proper message why the payment fails for the amounts lower than minimal amount. Like something wrong either on the backend side and we get the wrong response or on our side

I reproduced that behavior (both successful $0.20 payment the second time in a row and lack of specific error message) on tag [21.2](https://github.com/woocommerce/woocommerce-android/tree/21.2) before payments refactor was merged into trunk, so it's definitely backend behavior. My bet would be it's because of the test mode in Woo Payments, but might be worth looking into it (creating issue [here](https://github.com/woocommerce/woocommerce-android/issues/13110)).

samiuelson committed:
Clean up tests after merge

samiuelson committed:
Crash in case order is null illegally
